Add fixed output length SHAKE algorithms.

slontis opened the pull request:
By adding fixed output algorithms such as SHAKE256-192 (which is 24 bytes), the SHAKE algorithm can then be used just like any other fixed sized digest, without needing to specify additional parameters. Some algorithms such as HSS/XMSS use fixed sized outputs, and this avoids code that then needed to special case SHAKE. These fixed size variants fail if you try to change the xoflen.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

slontis commented:
Looks like something broke on merge. Will have to fix next week.

t8m commented on doc/man3/EVP_DigestInit.pod:
```diff
@@ -488,8 +488,10 @@ EVP_MD_CTX_set_params() can be used with the following OSSL_PARAM keys:
 Sets the digest length for extendable output functions.
 The value should not exceed what can be given using a B<size_t>.
-It may be used by BLAKE2B-512, SHAKE-128 and SHAKE-256 to set the
+It may be used by B<BLAKE2B-512>, B<SHAKE-128> and B<SHAKE-256> to set the
```
Ha, this is an obsolete artefact - BLAKE2B should not be mentioned here at all.

slontis commented:
Rebased.
ping

slontis commented:
@t8m I have added SHAKE-128/256 also.

Looks like these will also be useful in other places..
e.g. For signing with ECDSA and RSA-PSS (which use the default sizes of 32 and 64 for SHAKE 128 and 256).

I have changed the names section so that the OID's are associated with these fixed functions (since this is most likely the most common way the SHAKE would be used for DER encoding)

slontis commented:
ping

t8m commented:
> @t8m I have added SHAKE-128/256 also.

I do not see it in the PR?

slontis commented:
Pushed :)

t8m commented on providers/implementations/include/prov/names.h:
```diff
@@ -231,8 +231,12 @@
 #define PROV_NAMES_KECCAK_384 "KECCAK-384"
 #define PROV_NAMES_KECCAK_512 "KECCAK-512"
 
-#define PROV_NAMES_SHAKE_128 "SHAKE-128:SHAKE128:2.16.840.1.101.3.4.2.11"
-#define PROV_NAMES_SHAKE_256 "SHAKE-256:SHAKE256:2.16.840.1.101.3.4.2.12"
```
Hmm... I am afraid this is an API break. I suspect there will be a breakage if you run this PR with an older FIPS provider and load both default and the old fips provider. There will be a naming mismatch or it will make the  PROV_NAMES_SHAKE_128_256 and PROV_NAMES_SHAKE_128 the same. Similarly for SHAKE-256 and SHAKE-256/512.

slontis replied:
It turns out that it is even worse, Moving the oid breaks even the running tests..
I assume in the name mapping that it maps the first name back to an oid? This sounds like a bad thing to be doing.
So when I try selecting SHAKE-128/256 it go tthe SHAKE-128 table entry.

So I have taken the OID change out of this PR. BUT this is eventually going to bite us at some point.. If the OID in a signature was SHAKE-128 and we tried interoping with something else we are going to get the wrong thing potentially if it assumes that the size is supposed to be 256 bit (due to the fact that our implementation incorrectly choses a default of 128 bits currently). We may need to do a breaking change to fix this..

t8m reviewed (commented):
Good work. Please add CHANGES.md entry and I'll approve.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 216 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 247 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 278 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 309 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 340 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 371 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 402 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 433 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 464 days ago

vdukhovni commented:
Does it make sense to keep this open? Or has it been overtaken by events?
What is the future of this PR?

@slontis

slontis commented:
Closing, the reasons why this is not a good way forward have been discussed many times now..
